File: app/src/main/java/com/example/kotlinbasic/practise/mvvm_kotlin/MainActivity.java
package com.example.kotlinbasic.practise.mvvm_kotlin;


import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;

import com.example.kotlinbasic.R;
import com.example.kotlinbasic.databinding.ActivityMvvmBinding;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // ViewModel updates the Model
        // after observing changes in the View

        // model will also update the view
        // via the ViewModel
        ActivityMvvmBinding activityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_mvvm);
        activityMainBinding.setViewModel(new AppViewModel());
        activityMainBinding.executePendingBindings();

    }

    @BindingAdapter({"toastMessage"})
    public static void runMe(View view, String message) {
        if (message != null)
            Toast.makeText(view.getContext(), message, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/kotlinbasic/practise/LiveData/MainActivity.kt
package com.example.kotlinbasic.practise.LiveData

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.example.kotlinbasic.R
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_counter)

        val viewmodel = ViewModelProvider(this).get(MainActivityViewModel::class.java)

        viewmodel.startCounter()

        viewmodel.seconds().observe(this, Observer {
            textView.text = it.toString()
        })

        viewmodel.finished().observe(this, Observer {
            textView.text ="Finished"
        })

    }
}